subsys/pm/pm.c

```diff
@@ -24,7 +24,17 @@ LOG_MODULE_REGISTER(pm, CONFIG_PM_LOG_LEVEL);
 
 static ATOMIC_DEFINE(z_post_ops_required, CONFIG_MP_NUM_CPUS);
 static sys_slist_t pm_notifiers = SYS_SLIST_STATIC_INIT(&pm_notifiers);
-static struct pm_state_info z_power_states[CONFIG_MP_NUM_CPUS];
+
+/*
+ * Properly initialize cpu power states. Do not make assumptions that
+ * ACTIVE_STATE is 0
+ */
+#define PM_STATE_INFO_DEFINE(__, _)		\
+	{ PM_STATE_ACTIVE, 0, 0, 0 },
+static struct pm_state_info z_power_states[CONFIG_MP_NUM_CPUS] = {
+	UTIL_LISTIFY(CONFIG_MP_NUM_CPUS, PM_STATE_INFO_DEFINE)
+};
```

Comment:
I think this should be more in line with what is used for the states API:

```suggestion
#define CPU_STATE_INIT(node_id)		\
	{ .state = PM_STATE_ACTIVE },
static struct pm_state_info z_power_states[] = {
	DT_FOREACH_CHILD(DT_PATH(cpus), CPU_STATE_INIT)
};
```

Reply:
yeh, that makes sense. The problem (if I not mistaken) is that there are some targets with more cpus than the ones declared in DT. Or cases, were the target overrides the number of CPUs changing `CONFIG_MP_NUM_CPUS`.

Reply:
> yeh, that makes sense. The problem (if I not mistaken) is that there are some targets with more cpus than the ones declared in DT. Or cases, were the target overrides the number of CPUs changing `CONFIG_MP_NUM_CPUS`.

In such cases: they should declare all CPUs in DT, or use DT overlays to add/delete CPU nodes. If `CONFIG_MP_NUM_CPUS` can be different from the number of CPUs in DT, then we need to know which ones in DT should be handled by PM, now there's no such mechanism.

Reply:
Don't disagree. Let me take deeper look, right now the Kconfig option is the one that governor  the number of cpus and idle threads AFAIK. So, in case that option is bigger than the ones declared in DT the change will cause an overrun in that array. We may need to end up removing `CONFIG_MP_NUM_CPUS` to have implemented as you suggested.

Reply:
I've pushed a new version of this pr, renaming variables and applying some of the suggestions. It is still relying on `CONFIG_MP_NUM_CPUS`. The reason is that this options is used in many places, from samples to emulator options and runners.

This has to be discussed with others, the build option is much easier to use when testing than having it in DT, for example for tests where you don't know how many cores the target has but you want to use only 1. I'll move forward with this topic though.